source/Weather.cpp

```diff
@@ -0,0 +1,115 @@
+	: hazard(hazard), totalLifetime((totalLifetime > 0) ? totalLifetime : lifetime), lifetime(lifetime), strength(strength)
+{
+}
+
+
+
+const Hazard *Weather::GetHazard() const
+{
+	return hazard;
+}
+
+
+
+bool Weather::HasWeapon() const
+{
+	return hazard->IsWeapon();
+}
+
+
+
+int Weather::Period() const
+{
+	// If a hazard deviates, then the period is divided by the square root of the
+	// strength. This is so that as the strength of a hazard increases, it gets both
+	// more likely to impact the ships in the system and each impact hits harder.
+	if(hazard->Deviates())
+		return max(1, static_cast<int>(static_cast<double>(hazard->Period()) / sqrt(Strength())));
+	else
+		return hazard->Period();
+}
+
+
+
+double Weather::DamageMultiplier() const
+{
+	// If a hazard deviates, then the damage is multiplied by the square root of the
+	// strength. This is so that as the strength of a hazard increases, it gets both
+	// more likely to impact the ships in the system and each impact hits harder.
+	if(hazard->Deviates())
+	{
+		// If the square root of the strength is greater than the period, then Period()
+		// will return 1. Given this, we need to multiply the amount of strength
+		// going toward the damage by some corrective factor. Figure out what the "true
+		// period" is (without it bottoming out at 1) and divide that with the current
+		// period in order to correctly scale the damage so that the DPS of the hazard
+		// will always scale properly with the strength.
+		// This also fixes some precision lost by the fact that the period is an integer.
+		double truePeriod = static_cast<double>(hazard->Period()) / sqrt(Strength());
+		double multiplier = static_cast<double>(Period()) / truePeriod;
+		return sqrt(Strength()) * multiplier;
+	}
+	else
+		return strength;
+}
+
+
+
+int Weather::Step(vector<Visual> &visuals)
+{
+	// Environmental effects are created by choosing a random angle and distance from
+	// the system center, then creating the effect there. If a hazard has no range,
+	// effects are only created out to the invisible fence.
+	double minRange = hazard->MinRange();
+	double maxRange = hazard->MaxRange();
+	if(!maxRange)
+		maxRange = 10000.;
+	for(const auto &effect : hazard->EnvironmentalEffects())
+		for(int i = 0; i < static_cast<int>(effect.second * Strength()); ++i)
+		{
+			Point angle = Angle::Random().Unit();
+			double magnitude = (maxRange - minRange) * sqrt(Random::Real());
+			Point pos = (minRange + magnitude) * angle;
+			visuals.emplace_back(*effect.first, pos, Point(), Angle::Random());
+		}
+	return --lifetime;
+}
+
+
+
+double Weather::Strength() const
+{
+	if(hazard->Deviates())
+	{
+		double castTotal = static_cast<double>(totalLifetime);
+		double castCurrent = static_cast<double>(lifetime);
+		// Using a deviation of totalLifetime / 4.3 causes the strength of the weather to start and end at about 10% the maximum.
+		double DEVIATION = castTotal / 4.3;
```

Comment:
```suggestion
		const double DEVIATION = castTotal / 4.3;
```

Also, why is the variable name all uppercase?

Reply:
This is a constant that should really only need to be calculated when the Weather is created, not every time this function is called, hence why I had it all caps.